File: frontend/src/lib/config/connectors.ts
import type { CandidateState, ConnectorState, SourceTool } from '$lib/types/connector';

export const CONNECTOR_POLL_MS = 10_000;

export const CONNECTOR_STATE_LABELS: Record<ConnectorState, string> = {
	idle: 'Not connected',
	live: 'Receiving',
	stale: 'Idle',
	paused: 'Paused'
};

export const CONNECTOR_STATE_DOT: Record<ConnectorState, string> = {
	idle: 'bg-muted-foreground',
	live: 'bg-success',
	stale: 'bg-warning',
	paused: 'bg-muted-foreground'
};

export const SOURCE_TOOL_LABELS: Record<SourceTool, string> = {
	proxy: 'Proxy',
	repeater: 'Repeater',
	other: 'Other'
};

export const CANDIDATE_STATES: CandidateState[] = ['new', 'queued', 'scanned', 'ignored'];

export const CANDIDATE_STATE_LABELS: Record<CandidateState, string> = {
	new: 'New',
	queued: 'Queued',
	scanned: 'Scanned',
	ignored: 'Ignored'
};

export const NOTICE_LABELS: Record<string, string> = {
	sensitive: 'Sensitive path',
	admin: 'Administrative interface',
	unseen_by_scans: 'Not found by any scan',
	new_params: 'Parameters not seen by scans',
	server_error: 'Server error',
	non_standard_method: 'Uncommon method',
	out_of_scope: 'Out of scope'
};

export const NOTICE_HELP: Record<string, string> = {
	sensitive: 'The path matches a pattern associated with sensitive files.',
	admin: 'The path matches an administrative or authentication surface.',
	unseen_by_scans: 'A scan covered this target and did not record this path.',
	new_params: 'A scan recorded this path with a different parameter set.',
	server_error: 'The server returned a 5xx response.',
	non_standard_method: 'The method is not GET, POST, HEAD or OPTIONS.',
	out_of_scope: 'A bug bounty program lists this host as out of scope.'
};

export const LOUD_NOTICES = new Set(['sensitive', 'admin', 'server_error', 'out_of_scope']);

export function noticeTone(notice: string): string {
	if (notice === 'out_of_scope') return 'text-destructive font-medium';
	if (notice === 'sensitive' || notice === 'server_error') return 'text-destructive';
	if (notice === 'admin') return 'text-warning';
	if (notice === 'unseen_by_scans' || notice === 'new_params') return 'text-info';
	return 'text-muted-foreground';
}

export const INGEST_PATH = '/api/v1/connectors/ingest';

export function ingestEndpoint(): string {
	if (typeof location === 'undefined') return INGEST_PATH;
	return `${location.origin}${INGEST_PATH}`;
}

File: shared/definitions/connectors.py
"""Connector vocabulary: states, source tools, notice kinds and actions."""

from __future__ import annotations

from enum import StrEnum

MAX_NAME = 80
MAX_CONNECTORS = 20
MAX_BATCH = 500
MAX_QUEUE = 5000
MAX_BODY_SAMPLE = 4000
MAX_CANDIDATE_SCAN = 200
STALE_MINUTES = 30
LIVE_MINUTES = 2


class ConnectorKind(StrEnum):
    BURP = "burp"


class ConnectorState(StrEnum):
    IDLE = "idle"
    LIVE = "live"
    STALE = "stale"
    PAUSED = "paused"


CONNECTOR_STATE_LABELS: dict[str, str] = {
    ConnectorState.IDLE.value: "Not connected",
    ConnectorState.LIVE.value: "Receiving",
    ConnectorState.STALE.value: "Idle",
    ConnectorState.PAUSED.value: "Paused",
}


class SourceTool(StrEnum):
    """The proxy tool that produced the request."""

    PROXY = "proxy"
    REPEATER = "repeater"
    OTHER = "other"


SOURCE_TOOL_LABELS: dict[str, str] = {
    SourceTool.PROXY.value: "Proxy",
    SourceTool.REPEATER.value: "Repeater",
    SourceTool.OTHER.value: "Other",
}

INGESTED_TOOLS: frozenset[str] = frozenset(
    {SourceTool.PROXY.value, SourceTool.REPEATER.value}
)

MAX_PICKER_TARGETS = 500
MAX_SCOPE_HOSTS = 500
MAX_NOTICE_BATCH = 25
MANUAL_RUN_LABEL = "Manual testing"
BROWSING_RUN_LABEL = "Browsing"


class ActionKind(StrEnum):
    """Work handed back to the proxy."""

    REPEATER = "repeater"


ACTION_KIND_LABELS: dict[str, str] = {
    ActionKind.REPEATER.value: "Send to Repeater",
}

MAX_PENDING_ACTIONS = 200
MAX_ACTION_BATCH = 50
ACTION_TTL_MINUTES = 60


class CandidateState(StrEnum):
    NEW = "new"
    QUEUED = "queued"
    SCANNED = "scanned"
    IGNORED = "ignored"


CANDIDATE_STATE_LABELS: dict[str, str] = {
    CandidateState.NEW.value: "New",
    CandidateState.QUEUED.value: "Queued",
    CandidateState.SCANNED.value: "Scanned",
    CandidateState.IGNORED.value: "Ignored",
}


class NoticeKind(StrEnum):
    """Why a shape was flagged."""

    SENSITIVE = "sensitive"
    ADMIN = "admin"
    UNSEEN_BY_SCANS = "unseen_by_scans"
    NEW_PARAMS = "new_params"
    SERVER_ERROR = "server_error"
    NON_STANDARD_METHOD = "non_standard_method"
    OUT_OF_SCOPE = "out_of_scope"


NOTICE_LABELS: dict[str, str] = {
    NoticeKind.SENSITIVE.value: "Sensitive path",
    NoticeKind.ADMIN.value: "Administrative interface",
    NoticeKind.UNSEEN_BY_SCANS.value: "Not found by any scan",
    NoticeKind.NEW_PARAMS.value: "Parameters not seen by scans",
    NoticeKind.SERVER_ERROR.value: "Server error",
    NoticeKind.NON_STANDARD_METHOD.value: "Uncommon method",
    NoticeKind.OUT_OF_SCOPE.value: "Out of scope",
}

NOTICE_HELP: dict[str, str] = {
    NoticeKind.SENSITIVE.value: "The path matches a pattern associated with sensitive files.",
    NoticeKind.ADMIN.value: "The path matches an administrative or authentication surface.",
    NoticeKind.UNSEEN_BY_SCANS.value: "A scan covered this target and did not record this path.",
    NoticeKind.NEW_PARAMS.value: "A scan recorded this path with a different parameter set.",
    NoticeKind.SERVER_ERROR.value: "The server returned a 5xx response.",
    NoticeKind.NON_STANDARD_METHOD.value: "The method is not GET, POST, HEAD or OPTIONS.",
    NoticeKind.OUT_OF_SCOPE.value: "A bug bounty program lists this host as out of scope.",
}

# a notice in this set surfaces the row on its own
LOUD_NOTICES: frozenset[str] = frozenset(
    {
        NoticeKind.SENSITIVE.value,
        NoticeKind.ADMIN.value,
        NoticeKind.SERVER_ERROR.value,
        NoticeKind.OUT_OF_SCOPE.value,
    }
)

NOTICE_ORDER: tuple[str, ...] = (
    NoticeKind.OUT_OF_SCOPE.value,
    NoticeKind.SENSITIVE.value,
    NoticeKind.SERVER_ERROR.value,
    NoticeKind.ADMIN.value,
)

SAFE_METHODS: frozenset[str] = frozenset({"GET", "HEAD", "OPTIONS"})
COMMON_METHODS: frozenset[str] = frozenset(
    {"GET", "POST", "HEAD", "OPTIONS", "PUT", "PATCH", "DELETE"}
)

AUTH_HEADERS: tuple[str, ...] = (
    "authorization",
    "cookie",
    "x-api-key",
    "x-auth-token",
    "x-csrf-token",
)


def state_for(minutes_since: float | None, paused: bool) -> str:
    if paused:
        return ConnectorState.PAUSED.value
    if minutes_since is None:
        return ConnectorState.IDLE.value
    if minutes_since <= LIVE_MINUTES:
        return ConnectorState.LIVE.value
    if minutes_since <= STALE_MINUTES:
        return ConnectorState.STALE.value
    return ConnectorState.IDLE.value
